File: cpp/medium/findMaximumXOR.cpp
#include <unordered_set>
#include <vector>
using namespace std;

static const int maxBit = 30;
int findMaximumXOR(vector<int> &nums) {
  int rs = 0;
  unordered_set<int> set;
  for (int i = maxBit; i >= 0; i--) {
    set.clear();
    for (auto num : nums) {
      set.insert(num >> i);
    }
    rs = (rs << 1) + 1;
    bool flag = false;
    for (auto num : nums) {
      if (set.count((num >> i) ^ rs)) {
        flag = true;
        break;
      }
    }

    if (!flag) {
      rs--;
    }
  }
  return rs;
}

File: go/middle/findMaximumXOR.go
package main

func findMaximumXOR(nums []int) int {
	rs := 0
	for i := 30; i >= 0; i-- {
		set := make(map[int]bool)
		rs = (rs << 1) + 1
		flag := false
		for _, v := range nums {
			set[v>>i] = true
			if _, e := set[rs^(v>>i)]; e {
				flag = true
				break
			}
		}
		if !flag {
			rs--
		}
	}
	return rs
}
